Backport docs for System.Numerics.Vectors

carlossanlop opened the pull request:
Backporting public API documentation from MS Docs into triple slash comments for the System.Numerics.Vectors namespace.

This namespace was interesting for this experiment because all the source files live in `System.Private.CoreLib`, so the porting tool had to make sure to read the `System.Numerics.Vectors.csproj` dependent projects and find the source code files in them.

The full description of the purpose of this PR is described in the issue [#44969 - Pilot a new process that extracts API documentation from source code](https://github.com/dotnet/runtime/issues/44969) under the section "Bring documentation from Docs to triple slash".

I ran the [porting tool](https://github.com/carlossanlop/DocsPortingTool) using this command:

```
DocsPortingTool.exe \
-CsProj D:\runtime\src\libraries\System.Numerics.Vectors\src\System.Numerics.Vectors.csproj \
-Docs D:\dotnet-api-docs\xml \
-Save true \
-Direction ToTripleSlash \
-IncludedAssemblies System.Numerics \
-IncludedNamespaces System.Numerics
```

Please compare the ported docs with their source files in dotnet-api-docs: 
https://github.com/dotnet/dotnet-api-docs/tree/master/xml/System.Numerics

carlossanlop committed:
Unavoidable refresh by Visual Studio of the csproj file.

carlossanlop committed:
Backport Matrix3x2

carlossanlop committed:
Backport Matrix4x4

carlossanlop committed:
Backport Plane

carlossanlop committed:
Backport Quaternion

carlossanlop committed:
Backport Vector

carlossanlop committed:
Backport Vector2

carlossanlop committed:
Backport Vector3

carlossanlop committed:
Backport Vector4

carlossanlop committed:
Manually fix Plane &lt; and &gt;

carlossanlop committed:
Manually fix Vector wrong param names

carlossanlop committed:
Manually fix Vector2 &lt; and &gt;

carlossanlop committed:
Manually fix Vector3 &lt; and &gt;

carlossanlop committed:
Manually fix Vector4 &lt; and &gt;

carlossanlop committed:
Add GenerateDocumentationFile to csproj

carlossanlop commented:
Need to refresh this branch. There are some recent nullability changes causing a conflict.

safern commented on src/libraries/System.Numerics.Vectors/System.Numerics.Vectors.sln:
```diff
@@ -1,4 +1,8 @@
-﻿Microsoft Visual Studio Solution File, Format Version 12.00
```
I guess these changes to the sln file are not intentional?

carlossanlop replied:
Visual Studio made those changes automatically (first commit in the PR).

pgovind replied:
I think these changes occur when a newer version of VS opens the sln. AFAIK, reverting these changes work, so that's what I've always done.

carlossanlop replied:
It's kind of annoying having to revert the changes every time I have to submit a new commit. @safern would I break something if I keep the sln changes?

carlossanlop replied:
I'll keep this unless it causes any issues.

safern commented on src/libraries/System.Numerics.Vectors/src/System.Numerics.Vectors.csproj:
```diff
@@ -3,8 +3,9 @@
     <IsPartialFacadeAssembly>true</IsPartialFacadeAssembly>
     <TargetFrameworks>$(NetCoreAppCurrent)</TargetFrameworks>
     <Nullable>enable</Nullable>
+    <GenerateDocumentationFile>true</GenerateDocumentationFile>
```
Is System.Private.CoreLib already generating a documentation file? We need it to generate it and disable warnings until we port all the documentation to System.Private.CoreLib.

carlossanlop replied:
I wanted to confirm this with you.

S.P.CoreLib does generate docs, but they are not yet made mandatory. I assume this change needs to be reverted, and the `GenerateDocumentationFile` change needs to be the last change we add to CoreLib, after all APIs that live there have their documentation backported.

safern replied:
I think having the `GenerateDocumentationFile=true` here is valuable as a future proof in the case where we add public API to this assembly and the implementation were to live here as opposed to System.Private.CoreLib.

We will have to add some infrastructure to merge the System.Private.CoreLib.xml file into the reference assemblies that expose those APIs xmls.

carlossanlop replied:
Okay thanks. If you think it's safe to keep it, then I'll keep it.

safern commented:
> This namespace was interesting for this experiment because all the source files live in System.Private.CoreLib, so the porting tool had to make sure to read the System.Numerics.Vectors.csproj dependent projects and find the source code files in them.

For implementation that lives in System.Private.CoreLib, we should run the tool with `RuntimeFlavor=Mono` so that specific API that is exposed via the Mono S.P.CoreLib if any, gets it's docs ported.

carlossanlop commented:
> we should run the tool with RuntimeFlavor=Mono

@safern are you saying I should execute the DocsPortingTool with this argument?

carlossanlop committed:
Revert csproj GenerateDocumentationFile change

safern commented:
> @safern are you saying I should execute the DocsPortingTool with this argument?

Yeah I'm saying the compilations when loading System.Private.CoreLib, should do a pass to get the locations with RuntimeFlavor=CoreClr (which is the default) and with RuntimeFlavor=Mono.

carlossanlop committed:
Revert file fully

jeffhandley commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Vector4.cs:
```diff
@@ -183,30 +197,33 @@ public static Vector4 UnitW
 
-        /// <summary>Multiplies a vector by the given scalar.</summary>
-        /// <param name="left">The source vector.</param>
+        /// <summary>Multiples the specified vector by the specified scalar value.</summary>
```
Typo: Should be "Multiplies"

carlossanlop replied:
Thanks for catching those. I think fixing them by hand is the best approach here. This wasn't a problem with the tool, but rather a real typo in the source.

jeffhandley commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Vector3.cs:
```diff
@@ -161,30 +171,33 @@ public static Vector3 UnitZ
 
-        /// <summary>Multiplies a vector by the given scalar.</summary>
-        /// <param name="left">The source vector.</param>
+        /// <summary>Multiples the specified vector by the specified scalar value.</summary>
```
Typo: Should be "Multiplies"

jeffhandley commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Vector3.cs:
```diff
@@ -161,30 +171,33 @@ public static Vector3 UnitZ
-        /// <summary>Multiplies a vector by the given scalar.</summary>
-        /// <param name="left">The scalar value.</param>
-        /// <param name="right">The source vector.</param>
+        /// <summary>Multiples the scalar value by the specified vector.</summary>
```
Typo: Should be "Multiplies"

jeffhandley commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Vector.cs:
```diff
@@ -900,46 +765,40 @@ public static Vector<double> Floor(Vector<double> value)
 
-        /// <summary>
-        /// Widens a Vector{Byte} into two Vector{UInt16}'s.
+        /// <summary>Widens a <c>Vector&lt;Byte&gt;</c> into two <c>Vector&lt;UInt16&gt;</c> instances.</summary>
```
There are a bunch of `&lt;` and `&gt;` contents below here. Should these be encoded differently?

carlossanlop replied:
Good catch. These should have been mentioned  as <`see cref="docid" />`.

carlossanlop replied:
Note: This is also a case of something that came as is from MS Docs. I'll fix it manually here, so next time we sent these docs to MS Docs, they'll get the correct content.

carlossanlop replied:
This particular case is referring to the type of the passed parameters, which is `Vector<byte>`. It seems I cannot mention them as a `see cref`, which is why these comments mention them this way.

I'll leave these specific cases unmodified, but I did find a few other similar cases in the same file that can be fixed.

tannergooding replied:
> It seems I cannot mention them as a see cref, which is why these comments were added this way.

You can, as `cref="Vector{Byte}" />`, the issue is that somewhere between Roslyn and msdocs processing the reference, this is reverted to simply `Vector{T}` and we lose the generic type information.

carlossanlop replied:
Yep, I found that option to be ugly. I prefer to keep the generic type, so the `<c></c>` seems to be the most appropriate workaround.

jeffhandley commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Vector2.cs:
```diff
@@ -136,30 +147,33 @@ public static Vector2 UnitY
 
-        /// <summary>Multiplies a vector by the given scalar.</summary>
-        /// <param name="left">The source vector.</param>
+        /// <summary>Multiples the specified vector by the specified scalar value.</summary>
```
Typo: Should be "Multiplies"

jeffhandley commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Vector2.cs:
```diff
@@ -136,30 +147,33 @@ public static Vector2 UnitY
-        /// <summary>Multiplies a vector by the given scalar.</summary>
-        /// <param name="left">The scalar value.</param>
-        /// <param name="right">The source vector.</param>
+        /// <summary>Multiples the scalar value by the specified vector.</summary>
```
Typo: Should be "Multiplies"

jeffhandley commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Vector4.cs:
```diff
@@ -183,30 +197,33 @@ public static Vector4 UnitW
-        /// <summary>Multiplies a vector by the given scalar.</summary>
-        /// <param name="left">The scalar value.</param>
-        /// <param name="right">The source vector.</param>
+        /// <summary>Multiples the scalar value by the specified vector.</summary>
```
Typo: Should be "Multiplies"

jeffhandley reviewed (commented):
There are several places with a typo of "Multiples" where it should be "Multiplies", as it was in the original `///` comments.

Should we fix the typos in `dotnet-api-docs`,  and re-run the doc port and update this PR? Or should we just fix those by-hand in this PR and wait for `dotnet-api-docs` to get rehydrated from the `///` comments after this is merged in?

carlossanlop committed:
Fix typo "Multiples" => "Multiplies"

carlossanlop committed:
Use <typeparamref name="T" /> instead of <c>T</c>

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs:
```diff
@@ -111,19 +123,20 @@ public Vector2 Translation
-        /// <summary>Multiplies two matrices together and returns the resulting matrix.</summary>
-        /// <param name="value1">The first source matrix.</param>
-        /// <param name="value2">The second source matrix.</param>
+        /// <summary>Returns the matrix that results from multiplying two matrices together.</summary>
```
I don't think we should block, but I'm really not a fane of this style of summary. I think `Multiples` first, as the action the method is perform, reads significantly better

carlossanlop replied:
@tannergooding if you can give me comment suggestions with the texts you prefer to see, that would be great. That way I can batch apply them.

tannergooding replied:
I think either `Multiplies two matrices together to compute the product` is simple, explanatory, and puts it in common math terms.

Likewise, `Divides two matrices to compute the quotient`, `Adds two matrices to compute the sum`, `Subtracts two matrices to compute the difference`.

And when the types differ: `Multiplies a matrix by a float to compute the product` or `Divides a matrix by a float to compute the quotient`, etc

tannergooding replied:
(Assuming, of course, the docs team doesn't see any issues here)

gewarren replied:
Tanner's suggestions sound great to me.

carlossanlop replied:
I addressed the ones I could find in Matrix3x2 and in Matrix4x4. If you find more, please help me by using the `` ```suggestion`` feature from GitHub.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs:
```diff
@@ -490,10 +506,10 @@ public static Matrix3x2 CreateTranslation(float xPosition, float yPosition)
-        /// <param name="matrix">The source matrix.</param>
-        /// <param name="result">The output matrix.</param>
-        /// <returns>True if the operation succeeded, False otherwise.</returns>
+        /// <summary>Inverts the specified matrix. The return value indicates whether the operation succeeded.</summary>
```
Same here, I think `Attempts` or `Tries` first makes it more immediately obvious that this is
1. A badly named API that should have been called `TryInvert`
2. Is easier to read and ensures the user understand the operation/pattern

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix4x4.cs:
```diff
@@ -27,55 +28,71 @@ public struct Matrix4x4 : IEquatable<Matrix4x4>
         );
 
-        /// <summary>Value at row 1, column 1 of the matrix.</summary>
+        /// <summary>The first element of the first row.</summary>
```
In this case, I think there is probably a good in between between what we had and what is here.

Matrices are in terms of columns and rows and so I'd think "element" may be initially confusing to some users.

carlossanlop replied:
As the area owner, feel free to provide a suggested text and I can apply it.

tannergooding replied:
I don't want to block this PR on it, since we aren't changing anything.

I think it might be good to see if we can get a best/recommended practices for documentation text though so we can try to:
1. Be overall consistent across .NET
2. Find something that works in cases where the terminology may end up differing

carlossanlop replied:
Not sure what the action item is here. But if you'd like us to improve the docs, feel free to provide `` ```suggestion`` comments so I can bulk apply them.

pgovind commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs:
```diff
@@ -52,19 +59,22 @@ public struct Matrix3x2 : IEquatable<Matrix3x2>
 
-        /// <summary>Returns the multiplicative identity matrix.</summary>
+        /// <summary>Gets the multiplicative identity matrix.</summary>
+        /// <value>The multiplicative identify matrix.</value>
```
What is this <value> tag? Is it new? Don't we use <return> here instead usually?

carlossanlop replied:
Properties need their value to be described. Here's the full explanation of this tag in the dotnet-api-docs Wiki:

https://github.com/dotnet/dotnet-api-docs/wiki/Property-Value

gewarren commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Plane.cs:
```diff
@@ -7,52 +7,52 @@
-        /// <param name="y">The Y-component of the normal.</param>
-        /// <param name="z">The Z-component of the normal.</param>
-        /// <param name="d">The distance of the Plane along its normal from the origin.</param>
+        /// <summary>Creates a <see cref="System.Numerics.Plane" /> object from the X, Y, and Z components of its normal, and its distance from the origin on that normal.</summary>
```
```suggestion
        /// <summary>Creates a <see cref="T:System.Numerics.Plane" /> object from the X, Y, and Z components of its normal, and its distance from the origin on that normal.</summary>
```

gewarren commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Quaternion.cs:
````diff
@@ -8,28 +8,32 @@
 {
     /// <summary>Represents a vector that is used to encode three-dimensional physical rotations.</summary>
+    /// <remarks>The <see cref="System.Numerics.Quaternion" /> structure is used to efficiently rotate an object about the (x,y,z) vector by the angle theta, where:
+    /// ```
````
Not sure about using \``` in XML. Also, it should have a code language.

carlossanlop replied:
Good catch. These are the cases that we need to manually move to a file in dotnet-api-docs, then link to it here.

carlossanlop replied:
Found a bunch of these and I have converted them either to remarks with format + cdata, or to `<c></c>`.

gewarren commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Vector.cs:
```diff
@@ -54,502 +47,416 @@ public static Vector<double> ConditionalSelect(Vector<long> condition, Vector<do
         /// <param name="left">The first vector to compare.</param>
         /// <param name="right">The second vector to compare.</param>
-        /// <returns>The resultant vector.</returns>
+        /// <returns>The resulting integral vector</returns>
```
```suggestion
        /// <returns>The resulting integral vector.</returns>
```

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs:
```diff
@@ -111,19 +123,20 @@ public Vector2 Translation
+        /// <param name="value1">The first matrix.</param>
+        /// <param name="value2">The second matrix.</param>
         /// <returns>The product matrix.</returns>
+        /// <remarks>The <see cref="System.Numerics.Matrix3x2.op_Multiply" /> method defines the operation of the multiplication operator for <see cref="System.Numerics.Matrix3x2" /> objects.</remarks>
```
Doesn't C# add these automatically? If not should it be updated to do so?

gewarren replied:
Oh yes, you are right. So this should be fine then. https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/xmldoc/cref-attribute

gewarren replied:
There's even a code analysis rule that enforces *not* setting a prefix - https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1200.

carlossanlop replied:
@gewarren There is a bug preventing me from indicating a method group in triple slash comments (the equivalent of `Overload:` as you mentioned). There is an open issue tracking this problem in the csharplang repo: https://github.com/dotnet/csharplang/issues/320

Would you mind helping me increase the priority/visibility of the csharplang issue? I am currently running the tool in other larger assemblies, and this is going to become a very common problem when porting types with lots of overloads (Process, FileStream are giving me a headache).

gewarren replied:
Done!

carlossanlop commented on src/libraries/System.Private.CoreLib/src/System/Numerics/Matrix3x2.cs:
```diff
@@ -7,7 +7,8 @@
 {
-    /// <summary>A structure encapsulating a 3x2 matrix.</summary>
+    /// <summary>Represents a 3x2 matrix.</summary>
+    /// <remarks>[!INCLUDE[vectors-are-rows-paragraph](~/includes/system-numerics-vectors-are-rows.md)]</remarks>
```
@gewarren that's a good point. In the first PR, I was asked to remove the CDATA since it was not providing any value in triple slash comments, was using a lot of unnecessary space and polluting the code visually.

So I have a question for you to help address this: When we send intellisense xmls to the Docs build system, is it smart enough to automatically wrap non-empty remarks (including remarks that say `To be added.`) with these tags?:

 `<format type="text/markdown><![CDATA[ ... ]]></format>` 

If not, then we have two options:

A) Convert markdown links like these to valid xml. Do you know what's the non-markdown syntax for a link to an md file? And do you know if the Docs build system will be able to read that correctly and render the md document properly?
B) Talk to the Docs dev team to request converting any remarks we send to them to markdown (if they don't do it yet).

Thoughts?

safern replied:
Another thing to consider is that the xmls that Roslyn will produce after the source of truth are comments in source code, we will be shipping that as part of the intellisense xml files right? How will this look when a customer hovers over an API on VS and VS takes the info for that API from comments with these includes or `[CDATA]`?

carlossanlop replied:
AFAIK, remarks do not show up in VS intellisense in any way.

gewarren replied:
> Do you know what's the non-markdown syntax for a link to an md file?

I don't know. And the thing is, you don't just want a link, you actually want to import the contents of the linked file.

> When we send intellisense xmls to the Docs build system, is it smart enough to automatically wrap non-empty remarks?

No, it does not. See for example [here](https://github.com/dotnet/dotnet-api-docs/blob/733b2cecdd45ab3ed5efee2f001d9b07730f06e1/xml/Microsoft.Extensions.Diagnostics.HealthChecks/HealthCheckPublisherOptions.xml#L117).

carlossanlop replied:
@gewarren to address this problem, we can keep the CDATA section for remarks that require to show an include.

gewarren replied:
That's what the machinelearning repo does for its linked code snippets. See https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.FastTree/TreeTrainersCatalog.cs#L30.

carlossanlop committed:
Matrix3x2 adjust remarks.

carlossanlop committed:
Quaternion address remarks.

carlossanlop committed:
Plane adjust remarks.

carlossanlop committed:
Vector4 adjust remarks.

carlossanlop committed:
Matrix4x4 adjust remarks.

carlossanlop committed:
Vector2 adjust remarks.

carlossanlop committed:
Vector3 adjust remarks.

carlossanlop committed:
Address tannergooding and gewarren suggestions.

carlossanlop commented:
@gewarren @tannergooding @pgovind I marked this draft as ready for review. I addressed all your comments and suggestions and made sure the remarks with special markdown data show their information properly. Can you please take a look?

carlossanlop committed:
Matrix4x4 tries to invert.

carlossanlop committed:
Bring back GenerateDocumentationFile in csproj

safern commented on src/libraries/System.Numerics.Vectors/System.Numerics.Vectors.sln:
```diff
@@ -26,30 +30,27 @@ EndProject
-		{875DAA4F-1BC0-4EE2-9A04-EDD2FC5C773F} = {7A13467C-8469-4562-B00F-C4D61E8A1855}
-		{BD5C16FE-CAAA-45CB-91F3-C66A2493C518} = {7A13467C-8469-4562-B00F-C4D61E8A1855}
+	GlobalSection(SharedMSBuildProjectFiles) = preSolution
+		..\System.Private.CoreLib\src\System.Private.CoreLib.Shared.projitems*{634a3b2b-09f5-4810-b630-ade4d36c47df}*SharedItemsImports = 5
```
I don't understand why this is being added? Did you change this on VS?

carlossanlop replied:
No, I didn't modify the sln file myself. Visual Studio did this. I'll revert the sln file change.

carlossanlop replied:
@safern reverted.

carlossanlop committed:
Revert sln change

carlossanlop commented:
Learnings from this PR:

- Overloads can't be shown in triple slash comments. This is a bug, and it is tracked by https://github.com/dotnet/csharplang/issues/320 .
- When remarks contain a hyperlink or an include to an example, the whole element needs to be transformed to markdown (wrapped by `<format type="text/markdown"><![CDATA[ ... ]]></format>`).

carlossanlop commented:
The two CI failures are unrelated to my change. I already tried re-running both legs 4 times and the failure is consistent.

One CI leg is failing with this:
```
ERROR: helix-scripts 20191204.1 has requirement azure-core==1.0.0, but you'll have azure-core 1.7.0 which is incompatible.
ERROR: helix-scripts 20191204.1 has requirement azure-storage-blob==12.0.0, but you'll have azure-storage-blob 12.5.0 which is incompatible.
ERROR: helix-scripts 20191204.1 has requirement certifi==2018.11.29, but you'll have certifi 2020.6.20 which is incompatible.
```

And the other one with this:
```
D:\workspace\_work\1\s\.packages\microsoft.dotnet.sharedframework.sdk\6.0.0-beta.21105.5\targets\sharedfx.targets(273,5): error : The following files are missing entries in the templated manifest: [D:\workspace\_work\1\s\src\installer\pkg\sfx\Microsoft.NETCore.App\Microsoft.NETCore.App.Runtime.sfxproj]
D:\workspace\_work\1\s\.packages\microsoft.dotnet.sharedframework.sdk\6.0.0-beta.21105.5\targets\sharedfx.targets(273,5): error : pgort140.dll. Add these file names with extensions to the 'PlatformManifestFileEntry' item group for the runtime pack and corresponding ref pack to include them in the platform manifest. [D:\workspace\_work\1\s\src\installer\pkg\sfx\Microsoft.NETCore.App\Microsoft.NETCore.App.Runtime.sfxproj]
##[error].packages\microsoft.dotnet.sharedframework.sdk\6.0.0-beta.21105.5\targets\sharedfx.targets(273,5): error : (NETCORE_ENGINEERING_TELEMETRY=Build) The following files are missing entries in the templated manifest:
pgort140.dll. Add these file names with extensions to the 'PlatformManifestFileEntry' item group for the runtime pack and corresponding ref pack to include them in the platform manifest.

Build FAILED.
```
